### ISSUE TITLE
Better Inform Package Authors About the Need for Name-spaced Output Build Extensions

### ISSUE THREAD

**Issue by corwinsheahan-wf** [dart-lang/build#1696]
## Relevant Information
* [x] Dart SDK Version (`2.0.0-dev.69.4`)
* [x] What package(s) from this repo you are using:
    * `build: "0.12.7+2"`
    * `build_config: "0.3.1+2"`
    * `build_resolvers: "0.2.1+1"`
    * `build_runner: "0.9.2"`
    * `build_runner_core: "0.2.2+2"`
    * `built_value_generator: "5.5.5"`
* [x] What builder(s) you are using (or writing yourself):
    * `build.yaml` file exists but is empty. Builders come from `build_redux` and `built_value_generator`. 

## Description
Hi There!

We're in the process of moving from transformers to builders as part of our migration to Dart 2. When running builds with an empty `build.yaml` on a repository, we got the following error:
```
[SEVERE] Conflicting outputs
Both built_value_generator|built_value and built_redux may output example/current/bar_mixin.g.dart. 
Potential outputs must be unique across all builders. See https://github.com/dart-
lang/build/blob/master/docs/faq.md#why-do-builders-need-unique-outputs
```
Those two packages (`built_value_generator` and `build_redux`) are using the same output build extension of `.g.dart` and operate on some of the same input files. According to the error link, it looks like the package authors are responsible for ensuring the the uniqueness of their build extensions:

> If multiple Builders are configured to (possibly) output the same file you can:
> ...
> * Contact the author of the Builder and ask that a more unique output extension is chosen.

Since this isn't well documented in the [build.yaml docs](https://github.com/dart-lang/build/blob/master/build_config/README.md), and in the [`Builder` docs](https://github.com/dart-lang/build/tree/master/build), it would be very easy for package authors to follow examples and write their output extensions as `.g.dart` (NOTE: `built_value_generator` released a new major version, `v6.0.0`, which changes the extension to `.built_value.g.part`), instead of a sensibly name-spaced value that will be unique to that package, causing potential pain for consumers of that package. Additionally, changing the build extension after the fact would be a breaking change.

Could this be better evangelized in the docs for the `Builder` class and/or the `build.yaml` spec, so as to help mitigate this potential pain point?

**Comment by matanlurey** [comment]
Does the source_gen `CombiningBuilder` help here @kevmoo @natebosch?

**Comment by natebosch** [comment]
Yes, once `built_redux` migrates to `SharedPartBuilder` and applies `CombiningBuilder` then it and `built_value_generator` can both, in effect, output to `.g.dart`.
